Add PGAdmin & SubQuery Node Hot-Reload

jorgecuesta opened the pull request:
This pull request introduces significant improvements to the local development environment and security practices:

### Enhanced Local Development:

- **PGAdmin Integration:** Introduced PGAdmin for easier database management through Docker Compose.
- **Hot Reloading:** Added `nodemon` for automatic code reloading during development, reducing manual restarts.
- **Simplified Scripts and Docker Configuration:** Streamlined scripts in package.json and optimized Docker Compose configuration for better developer experience.
- **Clearer Local Dev Instructions:** Updated Localdev.run.xml with guidance for running the project in IntelliJ IDEs.

### Strengthened Security:

- **Removed Environment Files:** Eliminated .env files for environment variables, enhancing security by storing them outside the codebase.
- **Environment Variable Configuration:** Updated Dockerfiles to use environment variables for configuration, improving maintainability and security.
- **Production Hot-Reload Warning:** Added a warning message in the `node-entrypoint.sh` script if hot-reload is enabled in a production environment.

### Additional Improvements:

- Updated `cosmjs` submodule.
- Added comments for future Dockerfile enhancements.
- Clarified changes to `--workers=` in docker-compose.yml.
- Refactored Docker files and scripts for better maintainability.

These changes streamline the development workflow, improve project security, and lay the groundwork for further enhancements.

jorgecuesta committed:
Add PGAdmin services and update scripts.

This commit introduces PGAdmin to the Docker compose configuration, making local development easier. It also includes changes in the package.json scripts to utilize Docker compose. Notably, the commit incorporates an updated 'docker:dev:local' script and a new ignore file for Docker.

bryanchriswhite commented on docker-compose.yml:
```diff
@@ -15,6 +15,22 @@ services:
+      PGADMIN_DEFAULT_PASSWORD: admin
+      PGADMIN_LISTEN_PORT: 5050
+    volumes:
+      - ./docker/postgres/servers.json:/pgadmin4/servers.json
```
Is `servers.json` a standard or convention that postgres or the docker image supports? Could you please add a brief comment and link to any relevant documenation?

jorgecuesta replied:
@bryanchriswhite this is something supported by pgadmin; [PGADMIN_SERVER_JSON_FILE](https://www.pgadmin.org/docs/pgadmin4/latest/container_deployment.html#environment-variables)

bryanchriswhite commented on docker-compose.yml:
```diff
@@ -15,6 +15,22 @@ services:
       timeout: 5s
       retries: 5
 
+  pgadmin:
```
:ok_hand: Nice! :sunglasses:

bryanchriswhite commented on .run/Localdev.run.xml:
```diff
@@ -0,0 +1,8 @@
+<component name="ProjectRunConfigurationManager">
```
#PUC ("please update comment") - could you please add a comment explaining what this is/does?

bryanchriswhite commented on .dockerignore:
:pray:

jorgecuesta commented:
Btw here are few reasons to use docker compose instead of docker-compose:

- Simplicity: If Docker Compose features are integrated into the Docker CLI, you don’t need to install two separate tools anymore. You just need one (docker) and can run all commands from it.

- Eases Distribution: Docker is universally recognized, so having Compose as a part of the main Docker CLI ensures that the Compose specification can be readily available wherever Docker is.

- Better Integration: Compose is part of the Docker CLI now, as 'docker compose', which means it will have top-level integration and will likely get new features and improvements quicker than docker-compose.

- Compatibility: Generally, all of your existing docker-compose command line flags and functionality should still work with docker compose, ensuring backward compatibility.

bryanchriswhite commented on package.json:
```diff
@@ -9,20 +9,22 @@
     "docker:clean": "yarn docker:stop && docker volume rm pocketdex_postgres_data",
     "docker:tunnel": "echo 'if public key auth fails the password is `proxypass`.' && ssh -o StrictHostKeyChecking=no -N -R 26657:localhost:26657 proxyuser@localhost -p 2222",
-    "dev": "subql codegen && subql build && docker-compose pull && docker-compose up --remove-orphans",
+    "docker:dev:local": "docker compose up -d postgres pgadmin graphql-engine",
+    "dev:docker": "subql codegen && subql build && docker compose pull && docker compose up --remove-orphans",
+    "dev:local": "NODE_ENV=develop node vendor/subql-cosmos/packages/node/bin/run -f=./ --db-schema=app --workers=1 --batch-size=10 --unfinalized-blocks=true",
```
There are some things here I think need reformulating but don't have the energy to articulate tonight. I will follow up in the morning.

jorgecuesta replied:
100% we need to go over these scripts once we reach a more stable product because otherwise they are a bit chaotic

bryanchriswhite replied:
I'm fine merging this as-is so long as the README is up-to-date. Deal? :handshake: 

_(Bonus points for opening an issue to clean up the scripts.)_

jorgecuesta replied:
> Do not leave for tomorrow what you can do today

So I simplified the scripts, cleaned up things a bit, and added `nodemon` for hot-reload. Is not perfect, there is still room for improvement around docker builds but you should not need to build it if do not change vendor or package.json

jorgecuesta committed:
Update Docker configuration and local dev instructions

The Docker configuration has been updated: path to the pgAdmin server configuration file has been corrected. In addition, Localdev.run.xml has been updated to include a comment providing guidance for users utilizing IntelliJ IDEs to run the project locally. The pgAdmin server configuration file has also been renamed for consistency.

jorgecuesta committed:
Add nodemon for hot-reloading and clean up scripts

This commit adds nodemon for hot-reloading during development, reducing the need for manual restarts. Scripts in package.json have been simplified and organized to improve developer experience. Docker compose file and dockerfiles have been updated to support these changes, and unnecessary docker scripts have been removed from package.json.

bryanchriswhite commented on .yarn/cache/ignore-by-default-npm-1.0.1-78ea10bc54-441509147b.zip:
I think we should add .yarn/cache to the .gitignore and not track these.

jorgecuesta replied:
I think the same of you, just follow what already was done on the repository previously, please check the repo, `.yarn/cache` already exists, and looks like something the vendor modules are doing too.

So if we want to remove them maybe we can open another PR to clean up and prevent them from being added again?

bryanchriswhite commented on README.md:
````diff
@@ -34,36 +34,28 @@ Types will need to be regenerated any time the graphql.schema is changed.
 
 Build & start:
 ```shell
+# Run this only if indexing poktroll localnet. This will allows subquery-node to connect with the poktroll validator
+# leave this open in a separated terminal
+yarn docker:tunnel
````
This will only work _after_ `yarn run docker:start` because the proxy service won't be listening until then. Let's **either**:

a. Move this step below the `yarn run docker:start` step.
b. #PUC to include this detail.

jorgecuesta replied:
This is working because if you check what that `yarn docker:tunnel` is doing on the update it also startup the proxy with `docker compose up -d proxy` so you can turn on this before anything else without issues and leave it running out there (if needed of course)

bryanchriswhite commented on docker-compose.yml:
```diff
@@ -37,6 +53,12 @@ services:
       context: .
       dockerfile: ./docker/node.dockerfile
+      args:
+        # NOTE: this only applies when WATCH is not true on environment variables
+        # allow switching between dotenv file
+        # set nothing or comment to use .env or set develop to use .env.develop
+        # you can create your own .env.<something> and specify <something> as NODE_ENV value.
```
:fire: :sunglasses: Nice change, great comment!

bryanchriswhite commented on docker-compose.yml:
```diff
@@ -47,13 +69,20 @@ services:
       - -f=/app
       - --db-schema=app
-      - --workers=4
+      - --workers=1
```
Why switch to 1 worker?

jorgecuesta replied:
For development (with hot-reload, which is by default true) spam 4 workers produce tons more logs which complicates the debugging.

bryanchriswhite replied:
#PUC, just add that as a comment here please. :+1:

bryanchriswhite commented on docker-compose.yml:
```diff
@@ -47,13 +69,20 @@ services:
+      NODE_ENV: develop
     volumes:
       - ./:/app
+      - ignore_dist:/app/dist
+      - ignore_types:/app/src/types
```
Very clever :rocket:

bryanchriswhite commented on docker/node.dockerfile:
```diff
@@ -16,8 +19,8 @@ RUN yarn run vendor:clean
 # Build pocketdex
-WORKDIR /app
-RUN yarn codegen && yarn build
+RUN #NODE_ENV=$NODE_ENV yarn run codegen
```
Am I interpreting this correctly as commented out? It seems like we shouldn't need it since `yarn build` runs `yarn codegen`.
```suggestion
```

jorgecuesta replied:
My bad you are right, that line should be removed, because the next one which is `RUN NODE_ENV=$NODE_ENV yarn run build` does both `codegen` & `build`

bryanchriswhite commented on package.json:
```diff
@@ -4,25 +4,22 @@
+    "docker:build": "docker compose build",
+    "docker:build:no-cache": "docker compose build --no-cache",
+    "docker:start": "docker compose pull && docker compose up --remove-orphans",
+    "docker:stop": "docker compose stop",
```
:ok_hand:

bryanchriswhite commented on scripts/node-entrypoint.sh:
```diff
@@ -27,5 +29,22 @@ psql -v ON_ERROR_STOP=1 \
+  cp /project.yaml /app
+  update_project
+  # run the main node
+  NODE_ENV=$NODE_ENV env node /vendor/subql-cosmos/packages/node/bin/run "$@"
```
Let's add tini back here. I think something like:
```suggestion
  NODE_ENV=$NODE_ENV /sbin/tini -- env node /vendor/subql-cosmos/packages/node/bin/run "$@"
```

jorgecuesta replied:
`tini` is on the `ENTRYPOINT ["/sbin/tini", "--", "/entrypoint.sh"]` so every command executes inside of `entrypoint.sh` are already running with `tini`

bryanchriswhite replied:
Ah, right - thanks for clarifying. :sweat_smile:

bryanchriswhite commented on scripts/node-entrypoint.sh:
```diff
@@ -27,5 +29,22 @@ psql -v ON_ERROR_STOP=1 \
-# run the main node
-env node /vendor/subql-cosmos/packages/node/bin/run "$@"
\ No newline at end of file
+if [ "$WATCH" = "true" ]
```
:fire: :fire: :fire:

bryanchriswhite commented on scripts/node-entrypoint.sh:
```diff
@@ -27,5 +29,22 @@ psql -v ON_ERROR_STOP=1 \
+if [ "$WATCH" = "true" ]
+then
+  # call the first command if WATCH is true
+  echo "WATCH is true. Running the application with nodemon..."
```
What do you think about adding another log line if `NODE_ENV` is "production" that says something like `WARN: Hot-reload is not recommended in production. Unset $WATCH to disable it`?

jorgecuesta replied:
1000% true, let me add it.

bryanchriswhite reviewed (changes requested):
This is :fire: :fire: :fire: !

I had a couple questions and a couple of suggestions but otherwise I'm excited to see these DevEx improvements first-hand! :mechanical_arm: :sunglasses:

jorgecuesta committed:
Add warnings for hot-reload in production and improve Dockerfile

Added a condition in the node-entrypoint.sh script to issue a warning when hot-reload is activated in a production environment. In Dockerfile, grouped the codegen and build commands into a single run instruction to optimize the creation of Docker layers. Also added comments for future improvements regarding file copying and user permissions in Dockerfile.

jorgecuesta committed:
chore: update cosmjs submodule

bryanchriswhite reviewed (approved):
:rocket: :rocket: :rocket:

jorgecuesta committed:
Add a comment about the change to --workers= on docker-compose.yml

jorgecuesta committed:
Refactor Docker files and scripts, update package.json and remove environment files

This commit includes a major refactoring of Docker files and various scripts. The package.json file was updated with new script commands and improvements. The .env files, which previously held environment variables, were deleted for better security and replaced with a .env.sample file. Multiple .dockerfile scripts were updated to use environment variables for configuration. These changes will improve the maintainability of the project and the security of sensitive environment variables.

jorgecuesta commented:
@bryanchriswhite Based on our talks I pushed a major change to this PR, please if you have time give it a try before approving.

jorgecuesta committed:
Update Docker settings and script configurations

This commit makes several updates to Docker configurations and various script files. Changes include removing a line from .dockerignore, altering the handling of environment variables, updating permissions, and adjusting paths in several files. Additionally, the documentation is updated on how to modify the `.env` file. The purpose of these changes is to enhance the functionality and improve the integrity of the environment.

bryanchriswhite commented:
I'm having trouble building the composition. I get the same error whether I use `yarn run docker:build` or `yarn run docker:build:no-cache`:

![image](https://github.com/bryanchriswhite/pocketdex/assets/600733/d0ec4318-47a5-4d06-af23-e23bd73b85da)

jorgecuesta commented:
> I'm having trouble building the composition. I get the same error whether I use `yarn run docker:build` or `yarn run docker:build:no-cache`:
> 
> ![image](https://private-user-images.githubusercontent.com/600733/344393028-d0ec4318-47a5-4d06-af23-e23bd73b85da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2NzkyMTMsIm5iZiI6MTcxOTY3ODkxMywicGF0aCI6Ii82MDA3MzMvMzQ0MzkzMDI4LWQwZWM0MzE4LTQ3YTUtNGQwNi1hZjIzLWUyM2JkNzNiODVkYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyOVQxNjM1MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYjQ2YjVmYzcxYzZlZTIwM2M0NzU1Y2M4YjFmMDc0ODA5ODExMzZjMzQ1MGRiZDE3NjE2MDRiNWNiY2U5NWMwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.wlRrjxbFmJIOFGI8HHJZkTwSez7VNyJAZPPZpUbEkVQ)

base on the logs, look like you forget to create the .env copy from the .env.sample, btw, the old .env & .env.develop are not needed anymore

bryanchriswhite reviewed (approved):
Please take a look at https://github.com/pokt-scan/pocketdex/pull/1. Otherwise this looks great! 💪🔥

bryanchriswhite committed:
chore: ensure .env file exists when building (#1)

* chore: ensure .env file exists when building
* refactor: use error_log function
